src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala

```diff
@@ -0,0 +1,461 @@
+import io.iohk.ethereum.network.p2p.EthereumMessageDecoder
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.NewBlock
+import io.iohk.ethereum.network.rlpx.AuthHandshaker
+import io.iohk.ethereum.network.rlpx.RLPxConnectionHandler.RLPxConfiguration
+import io.iohk.ethereum.network.{EtcPeerManagerActor, ForkResolver, KnownNodesManager, PeerEventBusActor, PeerManagerActor, ServerActor}
+import io.iohk.ethereum.nodebuilder.{PruningConfigBuilder, SecureRandomBuilder}
+import io.iohk.ethereum.sync.FastSyncItSpec.{FakePeer, IdentityUpdate, customTestCaseResourceM, updateStateAtBlock}
+import io.iohk.ethereum.utils.ServerStatus.Listening
+import io.iohk.ethereum.utils.{Config, NodeStatus, ServerStatus, VmConfig}
+import io.iohk.ethereum.vm.EvmConfig
+import io.iohk.ethereum.{Fixtures, Timeouts}
+import monix.eval.Task
+import monix.execution.Scheduler
+import org.scalatest.{Assertion, AsyncFlatSpec, BeforeAndAfter, Matchers}
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util.Try
+
+class FastSyncItSpec extends AsyncFlatSpec with Matchers with BeforeAndAfter {
+  implicit val testScheduler = Scheduler.fixedPool("test", 16)
+
+  "FastSync" should "should sync blockchain without state nodes" in customTestCaseResourceM(FakePeer.start3FakePeersRes()) {
+    case (peer1, peer2, peer3) =>
+      for {
+        _ <- peer2.importNBlocksToTheTopForm(peer2.getCurrentState(), 1000)(IdentityUpdate)
+        _ <- peer3.importNBlocksToTheTopForm(peer3.getCurrentState(), 1000)(IdentityUpdate)
+        _ <- peer1.connectToPeers(Set(peer2.node, peer3.node))
+        _ <- peer1.startFastSync()
+        _ <- peer1.waitForFastSyncFinish()
+      } yield {
+        assert(peer1.bl.getBestBlockNumber() == peer2.bl.getBestBlockNumber() - peer2.testSyncConfig.targetBlockOffset)
+        assert(peer1.bl.getBestBlockNumber() == peer3.bl.getBestBlockNumber() - peer3.testSyncConfig.targetBlockOffset)
+      }
+  }
+
+  it should "should sync blockchain with state nodes" in customTestCaseResourceM(FakePeer.start3FakePeersRes()) {
+    case (peer1, peer2, peer3) =>
+      for {
+        _ <- peer2.importNBlocksToTheTopForm(peer2.getCurrentState(), 1000)(updateStateAtBlock(500))
+        _ <- peer3.importNBlocksToTheTopForm(peer3.getCurrentState(), 1000)(updateStateAtBlock(500))
+        _ <- peer1.connectToPeers(Set(peer2.node, peer3.node))
+        _ <- peer1.startFastSync()
+        _ <- peer1.waitForFastSyncFinish()
+      } yield {
+        val trie = peer1.getBestBlockTrie()
+        assert(peer1.bl.getBestBlockNumber() == peer2.bl.getBestBlockNumber() - peer2.testSyncConfig.targetBlockOffset)
+        assert(peer1.bl.getBestBlockNumber() == peer3.bl.getBestBlockNumber() - peer3.testSyncConfig.targetBlockOffset)
+        assert(trie.isDefined)
+      }
+  }
+
+
+  it should "should update target block" in customTestCaseResourceM(FakePeer.start2FakePeersRes()) {
```

Comment:
so i agree that test is can be a little bit not deterministic, but there is kinda no way around it. But the chances are highly stacked in favour of peer2 finishing first as:
- FastSync from one peer is slow, it goes like `headers -> bodies -> receipts -> headers ..` and even if header transaction root is root of empty trie we are doing full round trip and ask for body and receipts just to receive empty ones.
- When fast sync gets to target block it checks if peers block is still near it, and if not it update its target block. So if peer2 will manage to import 5blocks, during the time fast sync sync up 500, it will force it update it target block.